### ISSUE TITLE
Advanced queries

### ISSUE THREAD

**Issue by montogeek** [npms-io/npms-api#29]
Hi!

Great project!

I would like to ask about advanced queries, such as search by author or keywords as http://npmsearch.com/ do.

**Comment by mikeerickson** [comment]
In addition to @montogeek request, looking for the ability to search for multiple terms (ala npm install)

```
$http.get('https://api.npms.io/search', {
        json: true,
        query: {
          term: 'gulp-messenger gulp-phpunit gulp-phpspec',
        }
```

In addition, this could be expanded to support @montogeek request

```
$http.get('https://api.npms.io/search', {
        json: true,
        query: {
           author: 'mikeerickson',
           keyword: 'gulp'
        }
```

**Comment by satazor** [comment]
Support searching by keywords and author seems a good addition, we will definitively add those.

@mikeerickson we support multiple terms but they are used as AND and nor OR.

**Comment by satazor** [comment]
I'm currently looking into implementing this.

Regarding author:

- Author is defined by packageJson.author and can be a non-npm user
- Publisher is the last person that published/updated the module (npm user)
- Maintainer is all persons that have publish rights of the module (npm user)

If you link https://www.npmjs.com/~satazor, it shows all modules that I am qualified as a maintainer. When people are filtering by author what are they expecting? As you see author is a bit ambiguous. Suggestions?

**Comment by satazor** [comment]
//cc @sindresorhus

**Comment by sindresorhus** [comment]
Agreed. Should be the npm users with publish rights. `?maintainer=satazor` looks good.

**Comment by satazor** [comment]
@sindresorhus thanks for the feedback

The way I'm implementing it is not as a query parameter, but instead with `?q=query`, where query must obey a DSL. It looks similar to GitHub when you are searching for issues. Examples:

`maintainer:satazor exclude:deprecated keywords:windows cross-spawn`
`maintainer:sindresorhus exclude:insecure promise`
`score-effect:10 maintenance-weight:20 test runner`

This is a breaking change in the search API but I will support the old API until december. Having that said, it will support `maintainer:satazor` and also add `author and `publisher` support as I described above.

**Comment by satazor** [comment]
I've done some work and I got this stuff working locally. But since this is a breaking change I got to setup API versioning and ensure everything is done smoothly. Here's the API doc for the query DSL that allows advanced filtering and modifiers:

![screen shot 2016-09-17 at 13 38 52](https://cloud.githubusercontent.com/assets/1017236/18608242/1af4af28-7cdc-11e6-9689-88cc146c2ae2.png)

I expect to get this deployed in the next 4 days.
